README.md
# ispcr

[![PyPI](https://img.shields.io/pypi/v/ispcr?style=flat-square)](https://pypi.python.org/pypi/ispcr/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/ispcr?style=flat-square)](https://pypi.python.org/pypi/ispcr/)
[![PyPI - License](https://img.shields.io/pypi/l/ispcr?style=flat-square)](https://pypi.python.org/pypi/ispcr/)
[![Coookiecutter - Wolt](https://img.shields.io/badge/cookiecutter-Wolt-00c2e8?style=flat-square&logo=cookiecutter&logoColor=D4AA00&link=https://github.com/woltapp/wolt-python-package-cookiecutter)](https://github.com/woltapp/wolt-python-package-cookiecutter)


---

**Documentation**: [https://pommevilla.github.io/ispcr](https://pommevilla.github.io/ispcr)

**Source Code**: [https://github.com/pommevilla/ispcr](https://github.com/pommevilla/ispcr)

**PyPI**: [https://pypi.org/project/ispcr/](https://pypi.org/project/ispcr/)

---

A simple, light-weight package written in base Python to perform *in silico* PCR to determine primer performance.

**Currently in development**

## Installation

```sh
pip install ispcr
```
## Demonstration

### File-based *in silico* PCR

The main function to use in this package is `get_pcr_products`, which performs *in silico* PCR using two files:
  * `primer_file` - the path to fasta file containing your primers
    * This is currently limited to a fasta file containing two sequences, with the forward primer coming first and the reverse primer coming second
  * `sequence_file` the path to the fasta file containing the sequences to test your primers against

`get_pcr_products` will then iterate through the sequences in `sequence_file` and find all products amplified by the forward and reverse primer.

![](imgs/get_pcr_products_1.png)

### Refining results

You can also refine your results by using the `min_product_length` and `max_product_length` arguments, and only print out the columns you are interested in by using the `cols` argument. For example, if we're only interested in products between 100 and 250 bp and we only want the name of the sequence the product was amplified from, the length of the product, and the start and end position of the product in the sequence:

![](imgs/get_pcr_products_2.png)

### Writing out isPCR results to a file

`get_pcr_products` also takes an `output_file` argument. If provided, the results of the *in silico* PCR (including any product length restrictions or column selections) to that file. This will overwrite the file.

![](imgs/get_pcr_products_3.png)

### Sequence-based *in silico* PCR

The `get_pcr_products` function is a wrapper around `calculate_pcr_product`. The following arguments are required to run `calculate_pcr_product`:
  * `sequence`: the target sequence to test the primers against
  * `forward_primer`: the forward primer used to amplify the sequence
  * `reverse_primer`: the reverse primer (5'-3') used to amplify the sequence

`sequence`, `forward_primer`, and `reverse_primer` should be entered as `FastaSequence` objects. A `FastaSequence` is just a small convenience class to package a sequence with its header. An example run of `calculate_pcr_products` might look like:

![](imgs/calculate_pcr_product_1.png)

`calculate_pcr_product` uses all of the same arguments as `get_pcr_products`, so you can filter results and select columns just as before:

![](imgs/calculate_pcr_product_2.png)

This will also work with the `output_file` argument.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- GC content and melting temperature are now available

## [0.9.1] - 2023-01-03
### Changed
- Update documentation with new function names

## [0.9.0] - 2022-12-20
### Added
- Added a max_product_length option
- Add basic header support and product sequence name to output
- Added ability to select output columns with output string

## [0.6.0] - 2022-12-05
### Added
- find_pcr_product functionality with minimum product length

## [0.5.0] - 2022-12-02
### Added
- Initial product finding functionality

[Unreleased]: https://github.com/pommevilla/ispcr/compare/0.9.1...master
[0.9.1]: https://github.com/pommevilla/ispcr/compare/0.9.0...0.9.1
[0.9.0]: https://github.com/pommevilla/ispcr/compare/0.6.0...0.9.0
[0.6.0]: https://github.com/pommevilla/ispcr/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/pommevilla/ispcr/tree/0.5.0


tests/test_utils.py
from typing import Dict, Iterator, List

import pytest

from ispcr.FastaSequence import FastaSequence
from ispcr.utils import (
    InvalidColumnSelectionError,
    desired_product_size,
    filter_output_line,
    get_column_indices,
    get_invalid_bases,
    is_valid_cols_string,
    parse_selected_cols,
    read_fasta,
    reverse_complement,
)


class TestReverseComplement:
    def test_reverse_complement(self) -> None:
        test_string = "ACGT"
        expect_reverse_complement = "ACGT"
        actual_reverse_complement = reverse_complement(test_string)

        assert actual_reverse_complement == expect_reverse_complement

    def test_incorrect_bases(self) -> None:
        test_string = "AVGT"

        with pytest.raises(KeyError):
            reverse_complement(test_string)


class TestReaderUtils:
    @pytest.fixture(scope="class")
    def single_test_sequence(self) -> Iterator[List[FastaSequence]]:
        input_file = "tests/test_data/sequences/single_test.fa"
        fasta_sequences = []
        with open(input_file) as fin:
            for fasta_sequence in read_fasta(fin):
                fasta_sequences.append(fasta_sequence)
        yield fasta_sequences

    @pytest.fixture(scope="class")
    def multiple_test_sequences(self) -> Iterator[List[FastaSequence]]:
        input_file = "tests/test_data/sequences/met_r.fa"
        fasta_sequences = []
        with open(input_file) as fin:
            for fasta_sequence in read_fasta(fin):
                fasta_sequences.append(fasta_sequence)
        yield fasta_sequences

    def test_correct_number_of_entries(
        self, multiple_test_sequences: List[FastaSequence]
    ) -> None:
        assert len(multiple_test_sequences) == 20

    def test_read_single_fasta(self, single_test_sequence: FastaSequence) -> None:
        assert len(single_test_sequence) == 1

    # TODO: Mypy correctly identifies the single_test_sequence fixture as a string,
    # and throws an error when we try to use single_test_sequence.header.
    # PyTest runs fine, however. Hardcoding this for now to appease MyPy, but need to come
    # back and fix later.
    def test_correct_single_header(self, single_test_sequence: FastaSequence) -> None:
        expected_single_header = "single_test_sequence"
        single_test_sequence = FastaSequence("single_test_sequence", "GGG")
        actual_single_header = single_test_sequence.header

        assert expected_single_header == actual_single_header


class TestDesiredProductSize:
    def test_min_none_pass(self) -> None:
        min_product_length, max_product_length = None, 200
        potential_product_size = 150
        expected = True
        actual = desired_product_size(
            potential_product_size, min_product_length, max_product_length
        )

        assert actual == expected

    def test_min_none_fail(self) -> None:
        min_product_length, max_product_length = None, 200
        potential_product_size = 300
        expected = False
        actual = desired_product_size(
            potential_product_size, min_product_length, max_product_length
        )

        assert actual == expected

    def test_max_none_pass(self) -> None:
        min_product_length, max_product_length = 100, None
        potential_product_size = 150
        expected = True
        actual = desired_product_size(
            potential_product_size, min_product_length, max_product_length
        )

        assert actual == expected

    def test_max_none_fail(self) -> None:
        min_product_length, max_product_length = 100, None
        potential_product_size = 50
        expected = False
        actual = desired_product_size(
            potential_product_size, min_product_length, max_product_length
        )

        assert actual == expected

    def test_both_none_pass(self) -> None:
        min_product_length, max_product_length = None, None
        potential_product_sizes = [5, 50, 150, 250, 1000]
        expected = True

        for potential_product_size in potential_product_sizes:
            actual = desired_product_size(
                potential_product_size, min_product_length, max_product_length
            )
            assert actual == expected

    def test_both_pass(self) -> None:
        min_product_length, max_product_length = 75, 250
        potential_product_sizes = [75, 100, 150, 250]
        expected = True

        for potential_product_size in potential_product_sizes:
            actual = desired_product_size(
                potential_product_size, min_product_length, max_product_length
            )
            assert actual == expected

    def test_both_fail(self) -> None:
        min_product_length, max_product_length = 75, 250
        potential_product_sizes = [1, 50, 74, 251, 1000]
        expected = False

        for potential_product_size in potential_product_sizes:
            actual = desired_product_size(
                potential_product_size, min_product_length, max_product_length
            )
            assert actual == expected

    def test_improper_limits(self) -> None:
        min_product_length, max_product_length = 200, 100

        with pytest.raises(ValueError):
            desired_product_size(
                20,
                min_product_length=min_product_length,
                max_product_length=max_product_length,
            )


class TestIsValidHeaderString:
    @pytest.fixture(scope="class")
    def base_header(self) -> Iterator[str]:
        yield "fpri\trpri\tstart\tend\tlength\tpname\tpseq"

    def test_empty_string(self) -> None:
        expected = False
        actual = is_valid_cols_string("")

        assert actual == expected

    def test_existing_headers_in_order(self, base_header: str) -> None:

        for col_header in base_header.split():
            assert is_valid_cols_string(col_header)

    def test_existing_headers_out_of_order(self, base_header: str) -> None:
        from random import shuffle

        column_headers = base_header.split()
        shuffle(column_headers)

        for col_header in column_headers:
            assert is_valid_cols_string(col_header)

    def test_single_wrong_header(self) -> None:
        expected = False
        actual = is_valid_cols_string("forward_primer")

        assert actual == expected


class TestGetColumnIndicies:
    @pytest.fixture(scope="class")
    def base_header(self) -> Iterator[str]:
        yield "fpri\trpri\tstart\tend\tlength\tpname\tpseq"

    def test_empty_string(self) -> None:
        expected: List[int] = []
        actual = get_column_indices("")

        assert expected == actual

    def test_product_columns(self) -> None:
        expected = [5, 6]
        actual = get_column_indices("pname pseq")

        assert expected == actual

    def test_out_of_order(self) -> None:
        expected = [1, 0, 4, 3]
        actual = get_column_indices("rpri fpri length end")

        assert expected == actual

    def test_default_columns(self, base_header: str) -> None:
        expected = list(range(len(base_header.split())))
        actual = get_column_indices("fpri\trpri\tstart\tend\tlength\tpname\tpseq")

        assert actual == expected


class TestFilterOutputLine:
    @pytest.fixture(scope="class")
    def base_header(self) -> Iterator[str]:
        yield "fpri\trpri\tstart\tend\tlength\tpname\tpseq"

    @pytest.fixture(scope="class")
    def small_product_line(self) -> Iterator[str]:
        yield "forward_primer.f\treverse_primer.r\t0\t31\t31\ttest_sequence\tGGAGCATGCTATGTCGTAGCTGATGCAATTA"

    def test_no_output_columns_small_products(self, small_product_line: str) -> None:
        expected = ""
        col_indices: List[int] = []
        actual = filter_output_line(small_product_line, column_indices=col_indices)

        assert expected == actual

    def test_checks_for_default_header(self, base_header: str) -> None:
        expected = base_header
        actual = filter_output_line(base_header, list(range(7)))

        assert expected == actual

    def test_no_early_return_when_line_header_string_out_of_order(
        self, base_header: str
    ) -> None:
        expected = "pname\trpri\tend\tlength\tstart\tpseq\tfpri"
        actual = filter_output_line(base_header, [5, 1, 3, 4, 2, 6, 0])

        assert expected == actual

    def test_product_info_only(self, small_product_line: str) -> None:
        expected = "test_sequence\tGGAGCATGCTATGTCGTAGCTGATGCAATTA"
        actual = filter_output_line(small_product_line, [5, 6])

        assert expected == actual

    def test_returns_empty_string(self) -> None:
        expected = ""
        actual = filter_output_line("", list(range(7)))

        assert expected == actual


class TestParseSelectedCols:
    @pytest.fixture(scope="class")
    def base_header(self) -> Iterator[str]:
        yield "fpri\trpri\tstart\tend\tlength\tpname\tpseq"

    def test_exits_on_invalid_cols(self) -> None:
        with pytest.raises(InvalidColumnSelectionError):
            parse_selected_cols("invalid col string")

    def test_outputs_default_string(self, base_header: str) -> None:
        expected = list(range(len(base_header.split())))
        actual = parse_selected_cols(base_header)

        assert expected == actual


class TestGetInvalidBases:
    def test_simple_invalid_base_count(self) -> None:
        test_string = "ACGTV"
        expected = {"V": [4]}
        actual = get_invalid_bases(test_string)
        assert expected == actual

    def test_no_invalid_bases_found(self) -> None:
        test_string = "ACGT"
        expected: Dict[str, List[int]] = {}
        actual = get_invalid_bases(test_string)
        assert expected == actual

    def test_multiple_single_invalid_bases(self) -> None:
        test_string = "TZACFGH"
        expected = {"Z": [1], "F": [4], "H": [6]}
        actual = get_invalid_bases(test_string)
        assert expected == actual

    def test_multiple_times(self) -> None:
        test_string = "TZACFGH"
        expected = {"Z": [1], "F": [4], "H": [6]}
        actual = get_invalid_bases(test_string)
        assert expected == actual


tests/test_headers.py
from typing import Iterator, List

import pytest

from ispcr import calculate_pcr_product, get_pcr_products
from ispcr.FastaSequence import FastaSequence


class TestHeaders:
    @pytest.fixture(scope="class")
    def primers(self) -> Iterator[List[FastaSequence]]:
        forward_primer = FastaSequence("test_forward", "GGAG")
        reverse_primer = FastaSequence("test_reverse", "TAAT")
        yield [forward_primer, reverse_primer]

    @pytest.fixture(scope="class")
    def small_sequence_1(self) -> Iterator[FastaSequence]:
        sequence = FastaSequence("test_sequence", "GGAGCATGCTATGTCGTAGCTGATGCAATTA")
        yield sequence

    def test_no_header_calculate_pcr_product(
        self, primers: List[FastaSequence], small_sequence_1: FastaSequence
    ) -> None:
        expected_results = ""

        forward_primer, reverse_primer = primers
        pcr_results = calculate_pcr_product(
            sequence=small_sequence_1,
            forward_primer=forward_primer,
            reverse_primer=reverse_primer,
            header=False,
        ).splitlines()
        actual_results = "".join(
            [line for line in pcr_results if "forward_primer" in line]
        )

        assert expected_results == actual_results

    def test_basic_header_calculate_pcr_product(
        self, primers: List[FastaSequence], small_sequence_1: FastaSequence
    ) -> None:
        expected_results = "forward_primer\treverse_primer\tstart\tend\tlength\tproduct_name\tproduct_sequence"

        forward_primer, reverse_primer = primers
        pcr_results = calculate_pcr_product(
            sequence=small_sequence_1,
            forward_primer=forward_primer,
            reverse_primer=reverse_primer,
            header=True,
        ).splitlines()
        actual_results = "".join(
            [line for line in pcr_results if line == expected_results]
        )

        assert expected_results == actual_results

    def test_no_header_get_pcr_products(self) -> None:
        expected_results = ""

        pcr_results = get_pcr_products(
            primer_file="tests/test_data/primers/test_primers_1.fa",
            sequence_file="tests/test_data/sequences/small_sequence.fa",
            min_product_length=100,
            header=False,
        ).splitlines()
        actual_results = "".join(
            [line for line in pcr_results if "product_name" in line]
        )

        assert expected_results == actual_results

    def test_basic_header_get_pcr_products(self) -> None:
        expected_results = "forward_primer\treverse_primer\tstart\tend\tlength\tproduct_name\tproduct_sequence"

        pcr_results = get_pcr_products(
            primer_file="tests/test_data/primers/test_primers_1.fa",
            sequence_file="tests/test_data/sequences/small_sequence.fa",
            min_product_length=100,
            header=True,
        ).splitlines()
        actual_results = "".join(
            [line for line in pcr_results if line == expected_results]
        )

        assert expected_results == actual_results


tests/test_output_files.py
from os import listdir, remove
from os.path import exists
from typing import Iterator, List

import pytest

from ispcr import calculate_pcr_product, get_pcr_products
from ispcr.FastaSequence import FastaSequence


class TestCalculatePCRPRoduct:
    @pytest.fixture(scope="class")
    def primers(self) -> Iterator[List[FastaSequence]]:
        forward_primer = FastaSequence("test_forward", "GGAG")
        reverse_primer = FastaSequence("test_reverse", "TAAT")
        yield [forward_primer, reverse_primer]

    @pytest.fixture(scope="class")
    def small_sequence_1(self) -> Iterator[FastaSequence]:
        sequence = FastaSequence("test_sequence", "GGAGCATGCTATGTCGTAGCTGATGCAATTA")
        yield sequence

    def test_no_output_file_when_false(
        self, primers: List[FastaSequence], small_sequence_1: FastaSequence
    ) -> None:
        cwd_before = listdir()
        forward_primer, reverse_primer = primers
        _ = calculate_pcr_product(
            sequence=small_sequence_1,
            forward_primer=forward_primer,
            reverse_primer=reverse_primer,
            header=False,
            output_file=False,
        )
        cwd_after = listdir()

        assert cwd_before == cwd_after

    def test_output_file_with_name(
        self, primers: List[FastaSequence], small_sequence_1: FastaSequence
    ) -> None:
        output_file_name = "test.txt"
        forward_primer, reverse_primer = primers
        _ = calculate_pcr_product(
            sequence=small_sequence_1,
            forward_primer=forward_primer,
            reverse_primer=reverse_primer,
            header=False,
            output_file=output_file_name,
        )
        assert exists(output_file_name)
        remove(output_file_name)


class TestGetPCRProducts:
    def test_no_output_file_when_false(self) -> None:
        cwd_before = listdir()
        _ = get_pcr_products(
            primer_file="tests/test_data/primers/test_primers_1.fa",
            sequence_file="tests/test_data/sequences/small_sequence.fa",
            output_file=False,
        )
        cwd_after = listdir()

        assert cwd_before == cwd_after

    def test_output_file_with_name(self) -> None:
        output_file_name = "test.txt"
        _ = get_pcr_products(
            primer_file="tests/test_data/primers/test_primers_1.fa",
            sequence_file="tests/test_data/sequences/small_sequence.fa",
            output_file=output_file_name,
        )
        assert exists(output_file_name)
        remove(output_file_name)


src/ispcr/__init__.py
import re
from typing import Union

from ispcr.FastaSequence import FastaSequence
from ispcr.utils import (
    desired_product_size,
    filter_output_line,
    parse_selected_cols,
    read_sequences_from_file,
    reverse_complement,
)

BASE_HEADER = (
    "forward_primer\treverse_primer\tstart\tend\tlength\tproduct_name\tproduct_sequence"
)


def calculate_pcr_product(
    sequence: FastaSequence,
    forward_primer: FastaSequence,
    reverse_primer: FastaSequence,
    min_product_length: Union[int, None] = None,
    max_product_length: Union[int, None] = None,
    header: bool = True,
    cols: str = "all",
    output_file: Union[bool, str] = False,
) -> str:
    """Returns the products amplified by a pair of primers against a single sequence.

    This function is meant to be used by find_pcr_product, but if you'd like to use this separately,
    you must supply FastaSequence objects. A FastaSequence is just a convenient object to couple a header
    with a DNA string. For example,
        >>> forward_primer = FastaSequence("test_forward_primer", "ACTG")
        >>> reverse_primer = FastaSequence("test_reverse_primer", "ATTA")
        >>> target_sequence = FastaSequence("target_sequence", "ATGCTGATGCATGCTA")

    Inputs
    ------
    sequence: FastaSequence
        The fasta sequence to test for amplification.

    forward_primer: FastaSequence
        The forward primer to use.

    reverse_primer: FastaSequence
        The forward primer to use. Note that you should supply this

    min_product_length: None | int
        If provided, only return those products whose length are greater than or equal to this number.
        Defaults to None, which returns all products found.

    max_product_length: None | int
        If provided, only return those products whose length are less than or equal to this number.
        Defaults to None, which returns all products found.

    header: bool | str
        Whether or not to print a header on the results. Defaults to True. False will not print out the header.

    cols: str
        Which columns to print out. Defaults to "all," which prints out all the columns. A string can be supplied to
        only output the strings of interest. For example, cols="fpri rpri pname" will only output the names of the forward
        primer, reverse primer, and the target sequence when a target is found.
        Available options are:
            fpri - the name of the forward primer
            rpri - the name of the reverse primer
            start - the start location of the product in the target sequence
            end - the end location of the product in the target sequence
            length - the length of the product
            pname - the name of the sequence in which the target was found
            pseq - the nucleotide sequnce of the amplified product

    output_file: bool | str
        The file to write the results out to. Defaults to False, which will not print anything out. Providing a string
        will create that input file at that location. If set to True without providing a string, the output file
        will be of the form <DD-MM-YYYY_HH:MM:SS>.txt



    Outputs
    -------
    A tab-separated string containing all of the products amplified by the primers contained in the primer file.
    The fields are:
        1. Forward primer name
        2. Reverse primer name
        3. Start position of the product in the target sequence
        4. End position of the product in the target sequence
        5. Product length
        6. The product

    """

    products = []

    # Check cols string
    selected_column_indices = parse_selected_cols(cols)

    if header is True:
        products.append(filter_output_line(BASE_HEADER, selected_column_indices))

    forward_matches = [
        match.start()
        for match in re.finditer(forward_primer.sequence, sequence.sequence)
    ]

    for forward_match in forward_matches:
        tempseq = sequence[forward_match:]
        reverse_matches = [
            match.start()
            for match in re.finditer(
                reverse_complement(reverse_primer.sequence), tempseq
            )
        ]

        for reverse_match in reverse_matches:
            product = tempseq[: reverse_match + len(reverse_primer)]
            start = forward_match
            end = (
                forward_match + reverse_match + len(reverse_primer)
            )  # This is the end in the original sequence
            product_length = len(product)
            # if min_product_length is not None and product_length < min_product_length:
            #     continue
            if not desired_product_size(
                product_length, min_product_length, max_product_length
            ):
                continue

            product_line = f"{forward_primer.header}\t{reverse_primer.header}\t{start}\t{end}\t{product_length}\t{sequence.header}\t{product}"
            products.append(filter_output_line(product_line, selected_column_indices))

    results = "\n".join(products)

    if isinstance(output_file, str) is True:
        with open(output_file, "w") as fout:
            fout.write(results)

    return results


def get_pcr_products(
    primer_file: str,
    sequence_file: str,
    min_product_length: Union[int, None] = None,
    max_product_length: Union[int, None] = None,
    header: Union[bool, str] = True,
    cols: str = "all",
    output_file: Union[bool, str] = False,
) -> str:
    """Returns all the products amplified by a set of primers in all sequences in a fasta file.

    Inputs
    ------
    primer_file: str
        The path to the fasta file containing the primers to be tested. Currently, this primer
        file is expected to only contain two sequences, with the forward sequence appearing first.
        For an example:
            >test_1.f
            AGTCA
            >test_2.r
            TTATGC

    sequence_file: str
        The path to the fasta file containing the sequences to test the primers against.

    min_product_length: None | int
        If provided, only return those products whose length are greater than or equal to this number.
        Defaults to None, which returns all products found.

    max_product_length: None | int
        If provided, only return those products whose length are less than or equal to this number.
        Defaults to None, which returns all products found.

    header: bool
        Whether or not to print a header on the results. Defaults to True. False will not print out the header.

    cols: str
        Which columns to print out. Defaults to "all," which prints out all the columns. A string can be supplied to
        only output the strings of interest. For example, cols="fpri rpri pname" will only output the names of the forward
        primer, reverse primer, and the target sequence when a target is found.
        Available options are:
            fpri - the name of the forward primer
            rpri - the name of the reverse primer
            start - the start location of the product in the target sequence
            end - the end location of the product in the target sequence
            length - the length of the product
            pname - the name of the sequence in which the target was found
            pseq - the nucleotide sequnce of the amplified product

    output_file: bool | str
        The file to write the results out to. Defaults to False, which will not print anything out. Providing a string
        will create that input file at that location. If set to True without providing a string, the output file
        will be of the form <DD-MM-YYYY_HH:MM:SS>.txt


    Outputs
    -------
    A tab-separated string containing all of the products amplified by the primers contained in the primer file.
    The fields are:
        1. Forward primer name
        2. Reverse primer name
        3. Start position of the product in the target sequence
        4. End position of the product in the target sequence
        5. Product length
        6. The product

    """

    products = []

    # If anything gets passed for the header, it gets handled here instead of in
    # calculate_pcr_product.

    selected_column_indices = parse_selected_cols(cols)

    if header is True:
        products.append(filter_output_line(BASE_HEADER, selected_column_indices))

    primers = read_sequences_from_file(primer_file)
    forward_primer, reverse_primer = primers

    sequences = read_sequences_from_file(sequence_file)

    for sequence in sequences:
        new_products = calculate_pcr_product(
            sequence=sequence,
            forward_primer=forward_primer,
            reverse_primer=reverse_primer,
            min_product_length=min_product_length,
            max_product_length=max_product_length,
            header=False,
            cols=cols,
        )
        if new_products:
            products.append(new_products)

    results = "\n".join(products)

    if isinstance(output_file, str) is True:
        with open(output_file, "w") as fout:
            fout.write(results)

    return results


src/ispcr/utils.py
"""
This module contains various utilities used during in silico PCR.
"""

from typing import Dict, Iterator, List, TextIO, Union

from ispcr.errors import InvalidColumnSelectionError
from ispcr.FastaSequence import FastaSequence

BASE_HEADER = (
    "forward_primer\treverse_primer\tstart\tend\tlength\tproduct_name\tproduct_sequence"
)

COLUMN_HEADERS = {
    "fpri": 0,
    "rpri": 1,
    "start": 2,
    "end": 3,
    "length": 4,
    "pname": 5,
    "pseq": 6,
}
VALID_BASES = set("ACGT")


def desired_product_size(
    potential_product_length: int,
    min_product_length: Union[int, None] = None,
    max_product_length: Union[int, None] = None,
) -> bool:
    """Determines if a potential product's size is in the user's desired product range.

    Inputs
    potential_product_length - int
        The length of the potential product
    min_product_length: int | None
        The minimum product size the user will accept. If None, there is no lower limit.
    max_product_length: int | None
        The maximum product size the user will accept. If None, there is no lower limit.

    Outputs
    A boolean for whether the product length is between the min and max product length.

    Example
    desired_product_size(100, 75, 125)
    True
    desired_product_size(200, 75, 125)
    False
    """

    if min_product_length is None:
        if max_product_length is None:
            return True
        else:
            return potential_product_length <= max_product_length
    else:
        if max_product_length is None:
            return min_product_length <= potential_product_length
        else:
            if max_product_length < min_product_length:
                raise ValueError(
                    "min_product_length cannot be larger than max_product_length"
                )
            return min_product_length <= potential_product_length <= max_product_length


def reverse_complement(dna_string: str) -> str:
    """Returns the reverse complement of a DNA string.

    Inputs
    ------
    dna_string: str
        A string representing a DNA sequence. Supported bases are A, C, G, and T.

    Outputs
    -------
    The reverse complement of dna_string.

    Raises
    ------
    KeyError
        Raised if there is a base in dna_string that is not one of ACGT.

    Example
    -------
    >>> reverse_complement('GCTGA')
    'TCAGC'
    """

    complements = {"A": "T", "C": "G", "G": "C", "T": "A"}
    try:
        rev_seq = "".join([complements[s] for s in dna_string[::-1]])
    except KeyError as e:
        print(f"Base {e.args} not supported.")
        raise
    return rev_seq


def read_fasta(fasta_file: TextIO) -> Iterator[FastaSequence]:
    """An iterator for fasta files.

    Inputs
    ------
    fasta_file: TextIO
        An open file for reading

    Outputs
    -------
    An iterator yielding the the sequence names and sequences from a fasta file

    Example
    -------
    input_file = 'tests/test_data/sequences/met_r.fa.fasta'
    with open(input_file) as fin:
    for name, seq in read_fasta(fin):
        print(f'{name}\n{seq}')

    """
    name = None
    seq: List[str] = []
    for line in fasta_file:
        line = line.rstrip()
        if line.startswith(">"):
            if name:
                header = name
                sequence = "".join(seq)
                yield FastaSequence(header, sequence)
            name, seq = line[1:], []
        else:
            seq.append(line)
    if name:
        header = name
        sequence = "".join(seq)
        yield FastaSequence(header, sequence)


def read_sequences_from_file(primer_file: str) -> List[FastaSequence]:
    """
    Reads a fasta file, converts the sequences to FastaSequences, and returns them in a list.
    """
    sequences = []
    with open(primer_file) as fin:
        for fasta_sequence in read_fasta(fin):
            sequences.append(fasta_sequence)

    return sequences


def is_valid_cols_string(header_string: str) -> bool:
    """
    Internal helper to check if a column header string is valid.
    """
    if not header_string:
        return False

    for col_name in header_string.split():
        if col_name not in COLUMN_HEADERS:
            return False

    return True


def get_column_indices(header_string: str) -> List[int]:
    """
    Returns column indices based on a column header string.
    """
    return [COLUMN_HEADERS[col] for col in header_string.split()]


def filter_output_line(output_line: str, column_indices: List[int]) -> str:
    """
    Filters a single line of isPCR results based on selected column indices.
    """

    if not output_line.split():
        return ""
    elif column_indices == list(range(7)):
        return output_line
    else:
        columns = output_line.split()
        return "\t".join([columns[i] for i in column_indices])


def parse_selected_cols(cols: str) -> List[int]:
    """
    Returns a list of int indices based on a column header string.
    """
    if cols != "all":
        if not is_valid_cols_string(cols):
            raise InvalidColumnSelectionError("Invalid header string.")
        else:
            selected_column_indices = get_column_indices(cols)
    else:
        selected_column_indices = list(range(len(BASE_HEADER.split())))
    return selected_column_indices


def get_invalid_bases(seq: str) -> Dict[str, List[int]]:
    from collections import defaultdict

    invalid_bases = defaultdict(lambda: [])
    for i, base in enumerate(seq):
        if base not in VALID_BASES:
            invalid_bases[base].append(i)
    return invalid_bases
